internal/helm/repository.go

```diff
@@ -61,69 +61,84 @@ func NewChartRepository(repositoryURL string, providers getter.Providers, opts [
 	}
 
-	// Check for exact matches first
-	if len(version) != 0 {
-		for _, cv := range cvs {
-			if version == cv.Version {
-				return cv, nil
-			}
-		}
-	}
-
-	// Continue to look for a (semantic) version match
-	latestStable := len(version) == 0 || version == "*"
-	var match semver.Range
-	if !latestStable {
-		rng, err := semver.ParseRange(version)
+	// Parse version constraints if given (default strategy: latest)
+	var versionLatestStable = len(version) == 0 || version == "*"
+	var versionConstraints *semver.Constraints
+	if !versionLatestStable {
+		rng, err := semver.NewConstraint(version)
 		if err != nil {
 			return nil, err
 		}
-		match = rng
+		versionConstraints = rng
 	}
-	var filteredVersions semver.Versions
-	lookup := make(map[string]*repo.ChartVersion)
-	for _, cv := range cvs {
-		v, err := semver.ParseTolerant(cv.Version)
+
+	// Filter out chart versions that doesn't satisfy constraints if any,
+	// parse semver and build a lookup table
+	matchingVersions := make(semver.Collection, 0)
+	chartVersionLookup := make(map[*semver.Version]*repo.ChartVersion)
+	for _, chartVersion := range chartVersions {
+		// Check for exact matches first
+		if version == chartVersion.Version {
+			return chartVersion, nil
+		}
+
+		version, err := semver.NewVersion(chartVersion.Version)
 		if err != nil {
 			continue
 		}
-		// NB: given the entries are already sorted in LoadIndex,
-		// there is a high probability the first match would be
-		// the right match to return. However, due to the fact that
-		// we use a different semver package than Helm does, we still
-		// need to sort it by our own rules.
-		if match != nil && !match(v) {
+
+		if versionConstraints != nil && !versionConstraints.Check(version) {
 			continue
 		}
-		filteredVersions = append(filteredVersions, v)
-		lookup[v.String()] = cv
+
+		matchingVersions = append(matchingVersions, version)
+		chartVersionLookup[version] = chartVersion
 	}
-	if len(filteredVersions) == 0 {
+	if len(matchingVersions) == 0 {
 		return nil, fmt.Errorf("no chart version found for %s-%s", name, version)
 	}
-	sort.Sort(sort.Reverse(filteredVersions))
 
-	latest := filteredVersions[0]
-	if latestStable {
-		for _, v := range filteredVersions {
-			if len(v.Pre) == 0 {
-				latest = v
+	// Sort versions
+	sort.SliceStable(matchingVersions, func(i, j int) bool {
+		return !(func() bool {
+			left := matchingVersions[i]
+			right := matchingVersions[j]
+
+			if !left.Equal(right) {
+				return left.LessThan(right)
+			}
+
+			// Having chart creation timestamp at our disposal, we keep chronological
+			// order for packages with the same version, especially build metadata, since
+			// versions can be equal with different metadata, because it is not considered
+			// a part of the comparable version.
+			return chartVersionLookup[left].Created.Before(chartVersionLookup[right].Created)
+		})()
+	})
+
+	topMatchingVersion := matchingVersions[0]
+	if versionLatestStable {
```

Comment:
I kept the old logic for choosing a chart when the version is not specified here, however, it doesn't really live up to the promise in the `HelmChartSpec#Version` doc [here](https://github.com/fluxcd/source-controller/blob/main/api/v1beta1/helmchart_types.go#L34):
> Defaults to [the] **_latest_** when omitted.

In addition, taking the most recent chart (rather than the most recent stable chart) if the version is not specified is a better fit for the principle of least surprise. What'd you say, @hiddeco?

Anyway, it is a matter for another PR, but I would be happy to contribute a change.

Reply:
The spec contains a wrong explanation, as this logic is derived from Helm itself, and thus in line with what people would expect, see: https://github.com/helm/helm/blob/v3.3.4/pkg/repo/index.go#L152-L153